httpcore/_async/http11.py

```diff
@@ -168,12 +168,18 @@ async def _receive_response_body(self, request: Request) -> AsyncIterator[bytes]
             elif isinstance(event, (h11.EndOfMessage, h11.PAUSED)):
                 break
 
-    async def _receive_event(self, timeout: Optional[float] = None) -> H11Event:
+    async def _receive_event(
+        self, timeout: Optional[float] = None
+    ) -> Union[h11.Event, h11.PAUSED]:
         while True:
             with map_exceptions({h11.RemoteProtocolError: RemoteProtocolError}):
-                event = self._h11_state.next_event()
+                # The h11 type signature uses a private return type
+                event = cast(
+                    Union[h11.Event, h11.NEED_DATA, h11.PAUSED],
+                    self._h11_state.next_event(),
+                )
 
-            if event is h11.NEED_DATA:
+            if isinstance(event, h11.NEED_DATA):
```

Comment:
I worry how much this cast and isinstance will slow stuff down - as this is a hot loop

Reply:
we can hack around this by moving the cast outside the loop: https://github.com/encode/httpcore/blob/627db6cdcbf2d73a9b4c8800cb402b3bd6bf39df/httpcore/_async/http11.py#L30-L49
https://github.com/encode/httpcore/blob/627db6cdcbf2d73a9b4c8800cb402b3bd6bf39df/httpcore/_async/http11.py#L195-L207